// @jupyterlab/testutils/lib/index.js
"use strict";
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.waitForDialog = exports.testEmission = exports.sleep = exports.signalToPromises = exports.signalToPromise = exports.JupyterServer = exports.isFulfilled = exports.framePromise = exports.expectFailure = exports.dismissDialog = exports.dangerDialog = exports.acceptDialog = exports.FakeUserManager = exports.Mock = exports.defaultRenderMime = exports.NBTestUtils = exports.initNotebookContext = exports.DocumentWidgetOpenerMock = exports.createSession = exports.createFileContextWithKernel = exports.createFileContext = exports.createSessionContext = void 0;
var testutils_1 = require("@jupyterlab/apputils/lib/testutils");
Object.defineProperty(exports, "createSessionContext", { enumerable: true, get: function () { return testutils_1.createSessionContext; } });
var testutils_2 = require("@jupyterlab/docregistry/lib/testutils");
Object.defineProperty(exports, "createFileContext", { enumerable: true, get: function () { return testutils_2.createFileContext; } });
Object.defineProperty(exports, "createFileContextWithKernel", { enumerable: true, get: function () { return testutils_2.createFileContextWithKernel; } });
Object.defineProperty(exports, "createSession", { enumerable: true, get: function () { return testutils_2.createSession; } });
Object.defineProperty(exports, "DocumentWidgetOpenerMock", { enumerable: true, get: function () { return testutils_2.DocumentWidgetOpenerMock; } });
var testutils_3 = require("@jupyterlab/notebook/lib/testutils");
Object.defineProperty(exports, "initNotebookContext", { enumerable: true, get: function () { return testutils_3.initNotebookContext; } });
Object.defineProperty(exports, "NBTestUtils", { enumerable: true, get: function () { return testutils_3.NBTestUtils; } });
var testutils_4 = require("@jupyterlab/rendermime/lib/testutils");
Object.defineProperty(exports, "defaultRenderMime", { enumerable: true, get: function () { return testutils_4.defaultRenderMime; } });
exports.Mock = __importStar(require("./mock"));
var testutils_5 = require("@jupyterlab/services/lib/testutils");
Object.defineProperty(exports, "FakeUserManager", { enumerable: true, get: function () { return testutils_5.FakeUserManager; } });
var testing_1 = require("@jupyterlab/testing");
Object.defineProperty(exports, "acceptDialog", { enumerable: true, get: function () { return testing_1.acceptDialog; } });
Object.defineProperty(exports, "dangerDialog", { enumerable: true, get: function () { return testing_1.dangerDialog; } });
Object.defineProperty(exports, "dismissDialog", { enumerable: true, get: function () { return testing_1.dismissDialog; } });
Object.defineProperty(exports, "expectFailure", { enumerable: true, get: function () { return testing_1.expectFailure; } });
Object.defineProperty(exports, "framePromise", { enumerable: true, get: function () { return testing_1.framePromise; } });
Object.defineProperty(exports, "isFulfilled", { enumerable: true, get: function () { return testing_1.isFulfilled; } });
Object.defineProperty(exports, "JupyterServer", { enumerable: true, get: function () { return testing_1.JupyterServer; } });
Object.defineProperty(exports, "signalToPromise", { enumerable: true, get: function () { return testing_1.signalToPromise; } });
Object.defineProperty(exports, "signalToPromises", { enumerable: true, get: function () { return testing_1.signalToPromises; } });
Object.defineProperty(exports, "sleep", { enumerable: true, get: function () { return testing_1.sleep; } });
Object.defineProperty(exports, "testEmission", { enumerable: true, get: function () { return testing_1.testEmission; } });
Object.defineProperty(exports, "waitForDialog", { enumerable: true, get: function () { return testing_1.waitForDialog; } });
