

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>MacSyLib’s search engine &mdash; MacSyLib 1.0.3 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/graphviz.css?v=4ae1632d" />
      <link rel="stylesheet" type="text/css" href="../_static/_css/custom.css?v=b89f2dfc" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=baaebd52"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Frequently Asked Questions" href="FAQ.html" />
    <link rel="prev" title="Macromolecular models" href="model_definition.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            MacSyLib
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">User Guide</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="index.html#using-macsylib">Using MacSyLib</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="index.html#macsylib-functioning">MacSyLib functioning</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="model_definition.html">Macromolecular models</a></li>
<li class="toctree-l3"><a class="reference internal" href="model_definition.html#installing-models">Installing  models</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">MacSyLib’s search engine</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#functioning-overview">Functioning overview</a></li>
<li class="toctree-l4"><a class="reference internal" href="#a-searching-for-systems-components">A. Searching for Systems’ components</a></li>
<li class="toctree-l4"><a class="reference internal" href="#b-hits-browsing">B. Hits browsing</a></li>
<li class="toctree-l4"><a class="reference internal" href="#c-computing-candidate-systems-scores-ordered-mode">C. Computing candidate Systems’ scores (ordered mode)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#d-repeat-operations-b-and-c-for-the-other-models-being-searched">D. Repeat operations B and C for the other models being searched</a></li>
<li class="toctree-l4"><a class="reference internal" href="#e-computing-possible-solutions-defining-the-best-one-ordered-mode">E. Computing possible Solutions, defining the best one (ordered mode)</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="index.html#frequently-asked-questions">Frequently Asked Questions</a></li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modeler_guide/index.html">Modeller Guide</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide/index.html">Developer Guide</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">MacSyLib</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">User Guide</a></li>
      <li class="breadcrumb-item active">MacSyLib’s search engine</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="macsylib-s-search-engine">
<span id="functioning"></span><h1>MacSyLib’s search engine<a class="headerlink" href="#macsylib-s-search-engine" title="Link to this heading"></a></h1>
<section id="functioning-overview">
<h2>Functioning overview<a class="headerlink" href="#functioning-overview" title="Link to this heading"></a></h2>
<p>MacSyLib is able to use a variety of input files and options.
See <a class="reference internal" href="input.html#input-dataset-label"><span class="std std-ref">Input dataset</span></a> for more details. Below follows a description of its overall functioning.</p>
</section>
<section id="a-searching-for-systems-components">
<h2>A. Searching for Systems’ components<a class="headerlink" href="#a-searching-for-systems-components" title="Link to this heading"></a></h2>
<p>Initially, MacSyLib <strong>searches for the components</strong> of the <cite>System(s)</cite> to detect by sequence similarity search
(<code class="xref py py-func docutils literal notranslate"><span class="pre">macsylib.search_systems.search_systems()</span></code>).</p>
<p>1. From the list of <cite>System(s)</cite> to detect, a <strong>non-redundant list of components to search</strong> is built.
For each system, the list can include:</p>
<blockquote>
<div><ul class="simple">
<li><p>mandatory components</p></li>
<li><p>accessory components</p></li>
<li><p>neutral components</p></li>
<li><p>forbidden components</p></li>
<li><p>exchangeable components that can be functionally replaced by other components (usually by analogs or homologs).
These other components are thus also added to the list of components to search.</p></li>
</ul>
</div></blockquote>
<p>See <a class="reference internal" href="../modeler_guide/modeling.html#writing-models"><span class="std std-ref">here for more details on writing MacSyLib’s models.</span></a></p>
<p>2. HMMER is run on the corresponding set of components’ HMM profiles, and the hits are filtered according to the criteria defined
by the user or by default (see <a class="reference internal" href="input.html#hmmer-options"><span class="std std-ref">Hmmer options</span></a> and for more, the API <a class="reference internal" href="../developer_guide/api/report.html#hmmreport"><span class="std std-ref">report</span></a> object page).
This step, and the extraction of significant hits can be performed in parallel (<cite>worker</cite> config option).
See the <a class="reference internal" href="input.html#config-object-parameters"><span class="std std-ref">Config object</span></a>, and the <a class="reference internal" href="../developer_guide/api/search_genes.html#search-genes"><span class="std std-ref">search_genes API</span></a> for more details.</p>
<blockquote>
<div><a class="reference internal image-reference" href="../_images/msl_functionning_step1.svg"><img alt="../_images/msl_functionning_step1.svg" class="align-left" src="../_images/msl_functionning_step1.svg" style="height: 500px;" />
</a>
</div></blockquote>
</section>
<section id="b-hits-browsing">
<span id="system-assessment"></span><h2>B. Hits browsing<a class="headerlink" href="#b-hits-browsing" title="Link to this heading"></a></h2>
<p>The following steps depend on whether the input dataset is <strong>ordered</strong> (complete or nearly complete genome(s)),
or <strong>unordered</strong>  (metagenomes, or unassembled genome(s)) (see the <a class="reference internal" href="input.html#input-dataset-label"><span class="std std-ref">Input dataset</span></a> section).</p>
<p>In the case of <strong>ordered datasets</strong> (<cite>ordered_replicon</cite> or <cite>gembase</cite> search mode),
the hits are filtered to keep only hits related to the system’s model we are looking for.
These hits are used to build <strong>clusters of co-localized genes</strong> as defined <a class="reference internal" href="../modeler_guide/modeling.html#model-definition-genomic-orga"><span class="std std-ref">in the macsy-model files</span></a>.
These clusters are then screened to check for the model specifications such as the minimal quorum of
“Mandatory” or “Accessory” genes, or the absence of “Forbidden” components.</p>
<p>When the <strong>gene order is unknown</strong> (<cite>unordered</cite> search mode) the power of the analysis is more <strong>limited</strong>.
In this case, the presence of systems can only be suggested on the basis of
the <strong>quorum</strong> of components - and not based on genomic context information.</p>
<section id="for-ordered-datasets-building-clusters-of-components">
<h3>For <em>ordered</em> datasets: building clusters of components<a class="headerlink" href="#for-ordered-datasets-building-clusters-of-components" title="Link to this heading"></a></h3>
<p>The following two steps are reiterated for each model being searched.</p>
<ol class="arabic">
<li><p>The search starts with the filtering of hits to only keep the <strong>hits that are listed in the model</strong> (mandatory, accessory, neutral,
forbidden, exchangeable).</p></li>
<li><p>MacSyLib searches for sets of contiguous hits to build <strong>clusters</strong>, following the
<strong>(co-localization criterion)</strong> for each replicon, as defined in the MacSyLib’s model.
Two hits are deemed contiguous if their genomic location is separated by less than <em>d</em> protein-encoding genes, <em>d</em>
being the maximum of the two <cite>inter_gene_max_space</cite> parameters
from the two genes with hits (system-wise, or gene-specific parameter).
The <cite>loner</cite> components may form a cluster on their own.</p>
<a class="reference internal image-reference" href="../_images/msl_functionning_step2.svg"><img alt="../_images/msl_functionning_step2.svg" class="align-left" src="../_images/msl_functionning_step2.svg" style="height: 500px;" />
</a>
</li>
</ol>
<p>Once performed for each model searched, the <a class="reference internal" href="#combinatorial-exploration"><span class="std std-ref">next step</span></a> is performed.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The clusters that do not fulfill the quorum requirements are stored in the <a class="reference internal" href="outputs.html#rejected-candidates-txt"><span class="std std-ref">rejected_candidates.txt/tsv</span></a> file.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If several hits which co-locate have the same gene in the model. MSL does not consider them as a cluster.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If a group of gene which co-locate is composed solely of Neutral genes, It has not considered by MSL as a cluster.</p>
</div>
</section>
<section id="for-unordered-datasets">
<h3>For <em>unordered</em> datasets:<a class="headerlink" href="#for-unordered-datasets" title="Link to this heading"></a></h3>
<p>For each model being searched:</p>
<ol class="arabic simple">
<li><p>The Hits are filtered by model.</p></li>
<li><p>They are used to check if they reach the quorum (i.e., the clustering step is skipped as there is no notion of genetic distance in this search mode).</p></li>
<li><p>For each system, if the quorum is reached, hits are reported in the <a class="reference internal" href="outputs.html#all-systems-tsv-unordered"><span class="std std-ref">all_systems.tsv</span></a> output file.
It has to be noted that forbidden components are listed too, as they can also be informative for the user.</p></li>
</ol>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The “unordered” mode of detection is less powerful, as a single occurrence of a given model is filled for
an entire dataset with hits that origin is unknown. Please consider the assessment of systems with caution in this mode.</p>
</div>
<p>For unordered datasets, the <strong>search so ends</strong>, and you can generates the final <a class="reference internal" href="outputs.html#unordered-outputs"><span class="std std-ref">output files</span></a> with <code class="xref py py-mod docutils literal notranslate"><span class="pre">macsylib.io</span></code> module
(<code class="xref py py-func docutils literal notranslate"><span class="pre">macsylib.io.likely_systems_to_tsv()</span></code> and <code class="xref py py-func docutils literal notranslate"><span class="pre">macsylib.io.unlikely_systems_to_txt()</span></code>).</p>
</section>
</section>
<section id="c-computing-candidate-systems-scores-ordered-mode">
<span id="combinatorial-exploration"></span><h2>C. Computing candidate Systems’ scores (ordered mode)<a class="headerlink" href="#c-computing-candidate-systems-scores-ordered-mode" title="Link to this heading"></a></h2>
<p>This step only applies to the most powerful search mode, i.e., on <strong>ordered datasets</strong>.</p>
<p>The <strong>search engine</strong> explores the space of possible Solutions regarding the presence of Systems in replicons analysed.
It creates clusters of hits for Systems’ components separately for each System searched, and therefore might find
<strong>candidate occurrences of Systems that overlap</strong> in terms of components.
Moreover, if a System is possibly encoded at several locations on the replicon analysed (option <cite>multi_loci</cite> set to “True” in the model),
this calls for a <strong>combinatorial screening</strong> of the different clusters to assemble them into coherent systems regarding the macsy-models.</p>
<ul>
<li><p>For a given model, clusters are used to “fill up” Systems’ occurrence(s) according to the <strong>quorum criteria</strong>
defined in the System’s model (see methods <a class="reference internal" href="../developer_guide/api/system.html#macsylib.system.OrderedMatchMaker.match" title="macsylib.system.OrderedMatchMaker.match"><code class="xref py py-meth docutils literal notranslate"><span class="pre">macsylib.system.OrderedMatchMaker.match()</span></code></a>, <a class="reference internal" href="../developer_guide/api/system.html#macsylib.system.UnorderedMatchMaker.match" title="macsylib.system.UnorderedMatchMaker.match"><code class="xref py py-meth docutils literal notranslate"><span class="pre">macsylib.system.UnorderedMatchMaker.match()</span></code></a>):</p>
<blockquote>
<div><p>The <cite>min_genes_required</cite> and <cite>min_mandatory_genes_required</cite> thresholds must be reached.</p>
<blockquote>
<div><ul class="simple">
<li><p>In the case of the <cite>single-locus system</cite> search mode (default), each cluster in addition to potential loners
are evaluated for System’s assessment separately.</p></li>
<li><p>In the case of the <cite>multi-loci system</cite> search mode (<code class="docutils literal notranslate"><span class="pre">multi_loci=True</span></code>), each possible combination of clusters
is confronted to the quorum of the System being examined.</p></li>
</ul>
</div></blockquote>
<p>The sets of clusters that fulfill the quorum are reported as candidate Systems in the <cite>all_systems.txt</cite> and
<cite>all_systems.tsv</cite> output files (see <a class="reference internal" href="outputs.html#outputs"><span class="std std-ref">Output format</span></a>),
and they obtain a <strong>System’s score</strong> (see below).</p>
<p>The clusters that do not allow to form a candidate System are reported in the <cite>rejected_candidates.txt</cite> and  <cite>rejected_candidates.tsv</cite> output files.</p>
</div></blockquote>
</li>
<li><p>We introduce a <strong>scoring scheme for candidate Systems</strong>, to easily separate combinations of clusters that are readily
more similar to a system’s model than others.</p>
<p>The assumptions behind this scoring scheme are the following:</p>
<blockquote>
<div><ul>
<li><p>We set a score for the different types of genes/components when defining a <strong>cluster’s score</strong>. Here are the default values, but these <a class="reference internal" href="input.html#score-options"><span class="std std-ref">can be changed</span></a>:</p>
<blockquote>
<div><ul class="simple">
<li><p>+1.0 is added when a <cite>mandatory</cite> gene is present</p></li>
<li><p>+0.5 is added when an <cite>accessory</cite> gene is present</p></li>
<li><p>+0.0 is added when a <cite>neutral</cite> gene is present</p></li>
<li><p>*0.8 (a factor of 0.8) is applied to the above-scores when the function is fulfilled by an <cite>exchangeable</cite> gene</p></li>
<li><p>*0.7 (a factor of 0.7) is applied to the above-scores if the gene is a <cite>loner</cite> and <cite>multi system</cite> component.</p></li>
</ul>
</div></blockquote>
</li>
<li><p>When combinations of clusters are explored in order to fulfill macsy-models’ requirements and build candidate systems
(“multi_loci” mode, several clusters can make a complete <cite>System</cite>), we sum the score of clusters to assign a <cite>System</cite>’s score.</p></li>
<li><p>In addition, we want to <strong>favor concise sets of clusters</strong> to fulfill a <cite>System</cite>’s model.
We thus <strong>penalize the adjunction of a cluster</strong> to a candidate <cite>System</cite> when this cluster does not bring
any new components to the <cite>System</cite>’s quorum, or when it brings <strong>redundant components</strong>. Thus:</p>
<blockquote>
<div><ul class="simple">
<li><p>-1.5 is added when a <strong>redundant</strong> mandatory gene is added when adjuncting the cluster to a candidate <cite>System</cite></p></li>
<li><p>-1.5 is added when a <strong>redundant</strong> accessory gene is added when adjuncting the cluster to a candidate <cite>System</cite></p></li>
<li><p>for the components that are <cite>loner</cite> and <cite>multi system</cite>, the score of the loner component is added only if the function is not fulfilled in the other clusters.
In this case, even if there are several occurrences of the component, it is counted only once (and no penalty is applied).</p></li>
</ul>
</div></blockquote>
</li>
<li><p>Only candidate sets of clusters that fulfill a macsy-model and that are thus designated candidate <cite>Systems</cite>, obtain a <strong>System’s score</strong></p></li>
</ul>
</div></blockquote>
<p>In summary, a Systems’s score is made of two parts: the <strong>sum of the scores</strong> of the Clusters it is made of,
plus <strong>a penality part</strong> to avoid too much component’s redundancy in Cluster’s combinations.
The systems’ scoring step is exemplified in this figure:</p>
<a class="reference internal image-reference" href="../_images/msl_functionning_step3.svg"><img alt="../_images/msl_functionning_step3.svg" class="align-left" src="../_images/msl_functionning_step3.svg" style="height: 500px;" />
</a>
</li>
</ul>
</section>
<section id="d-repeat-operations-b-and-c-for-the-other-models-being-searched">
<h2>D. Repeat operations B and C for the other models being searched<a class="headerlink" href="#d-repeat-operations-b-and-c-for-the-other-models-being-searched" title="Link to this heading"></a></h2>
<a class="reference internal image-reference" href="../_images/msl_functionning_step4.svg"><img alt="../_images/msl_functionning_step4.svg" class="align-left" src="../_images/msl_functionning_step4.svg" style="height: 500px;" />
</a>
<p>This search for candidate <cite>Systems</cite> from different models results in a number of possible <cite>Solutions</cite>
representing combinations of putative sets of <cite>Systems</cite> in the analysed dataset.</p>
</section>
<section id="e-computing-possible-solutions-defining-the-best-one-ordered-mode">
<h2>E. Computing possible Solutions, defining the best one (ordered mode)<a class="headerlink" href="#e-computing-possible-solutions-defining-the-best-one-ordered-mode" title="Link to this heading"></a></h2>
<p>At the end of the previous step MacSyLib has computed all potential <cite>Systems</cite> present in the replicon,
made of combinations of Clusters and <cite>loner</cite> components that fulfill the model’s requirements,
which are themselves made of a subset of Hits (remember, Hits are at 1st filtered and treated separately for each model of System to be detected).
Candidate <cite>Systems</cite> may thus overlap by being partly made of the same components, or even partly being made of the same Clusters.</p>
<p>We define a <cite>Solution</cite> as being <strong>a set of compatible Systems</strong>, i.e. that do not have any overlaps between their components.
All possible <cite>Solutions</cite> are combinatorially explored and consist in all possible sets of compatible <cite>Systems</cite>.</p>
<p>A scoring scheme enables to separate between sets of <cite>Solutions</cite>. A <strong>Solution’s score</strong> is basically the <strong>sum of its Systems’ scores</strong>.
The overall procedure of exploring the space of all possible <cite>Solutions</cite> while finding the optimal one,
i.e. that with the maximal score, is performed at once using a graph solution to this problem, implemented in the <code class="docutils literal notranslate"><span class="pre">networkx</span> <span class="pre">package</span></code>.</p>
<p>We create a graph where each potential <cite>System</cite> is a vertex, and we create an edge between pairs of vertices
if they do not share any components (compatible <cite>Systems</cite>).
Once the graph is created we look for the <a class="reference external" href="https://en.wikipedia.org/wiki/Clique_problem#Definitions">maximum clique</a>
which maximizes the score. This allows to provide the user with one, or multiple <cite>Solutions</cite>
that have the <strong>best score possible</strong> among all combinations of compatible <cite>Systems</cite>.
All this step is performed by the <code class="xref py py-func docutils literal notranslate"><span class="pre">macsylib.solution.find_best_solution()</span></code></p>
<a class="reference internal image-reference" href="../_images/msl_functionning_step5.svg"><img alt="../_images/msl_functionning_step5.svg" class="align-left" src="../_images/msl_functionning_step5.svg" style="height: 500px;" />
</a>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="model_definition.html" class="btn btn-neutral float-left" title="Macromolecular models" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="FAQ.html" class="btn btn-neutral float-right" title="Frequently Asked Questions" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2014-2025, Bertrand Néron, Sophie Abby.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
  

<div class="licensing">
  <p>
    This documentation is licensed under <a href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License
      <img src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png">.
    </a>
  </p>
</div>


</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

<a class="github" href="https://github.com/gem-pasteur/macsylib">
    <img style="position: absolute; top: 0; right: 0; border: 0;"
         src="https://raw.githubusercontent.com/gem-pasteur/macsylib/refs/heads/main/doc/source/_static/github-banner.svg"
         alt="Fork me on GitHub" />
</a>



</body>
</html>