

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Input and Options of MacSyLib &mdash; MacSyLib 1.0.3 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/graphviz.css?v=4ae1632d" />
      <link rel="stylesheet" type="text/css" href="../_static/_css/custom.css?v=b89f2dfc" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=baaebd52"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Gembase format" href="gembase_convention.html" />
    <link rel="prev" title="MacSyLib Quick Start" href="quickstart.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            MacSyLib
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">User Guide</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="index.html#using-macsylib">Using MacSyLib</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="new.html">What’s new in MacSyLib</a></li>
<li class="toctree-l3"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html">MacSyLib Quick Start</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Input and Options of MacSyLib</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#input-dataset">Input dataset</a></li>
<li class="toctree-l4"><a class="reference internal" href="#config-object">Config object</a></li>
<li class="toctree-l4"><a class="reference internal" href="#configuration-file">Configuration file</a></li>
<li class="toctree-l4"><a class="reference internal" href="#in-house-input-files">In-house input files</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="outputs.html">Output format</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="index.html#macsylib-functioning">MacSyLib functioning</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#frequently-asked-questions">Frequently Asked Questions</a></li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modeler_guide/index.html">Modeller Guide</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide/index.html">Developer Guide</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">MacSyLib</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">User Guide</a></li>
      <li class="breadcrumb-item active">Input and Options of MacSyLib</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="input-and-options-of-macsylib">
<span id="input"></span><h1>Input and Options of MacSyLib<a class="headerlink" href="#input-and-options-of-macsylib" title="Link to this heading"></a></h1>
<section id="input-dataset">
<span id="input-dataset-label"></span><h2>Input dataset<a class="headerlink" href="#input-dataset" title="Link to this heading"></a></h2>
<p>The input dataset must be a set of protein sequences in <strong>Fasta format</strong> (see <a class="reference external" href="http://en.wikipedia.org/wiki/FASTA_format">http://en.wikipedia.org/wiki/FASTA_format</a>).
(The fasta file can be compressed in <em>gzip</em> format see note below)</p>
<p>The <a class="reference internal" href="#config-base-label"><span class="std std-ref">base section</span></a> in the configuration file (see <a class="reference internal" href="#config-definition-label"><span class="std std-ref">Configuration file</span></a>)
can be used to specify <strong>the path</strong> and the <strong>type of dataset</strong> to deal with,
as well as the <cite>sequence_db</cite> and <cite>db_type</cite> config options respectively,
described in the <a class="reference internal" href="#input-options"><span class="std std-ref">Input options</span></a>.</p>
<blockquote>
<div><p>Four types of protein datasets are supported:</p>
<blockquote>
<div><ul class="simple">
<li><p><em>unordered</em> : a set of sequences corresponding to a complete genome
(<em>e.g.</em> an unassembled complete genome)</p></li>
<li><p><em>ordered_replicon</em> : a set of sequences corresponding to an ordered complete replicon
(<em>e.g.</em> an assembled complete genome)</p></li>
<li><p><em>gembase</em> : a set of multiple ordered replicons, which format follows the convention described
in <a class="reference internal" href="gembase_convention.html#gembase-convention"><span class="std std-ref">Gembase format</span></a>.</p></li>
</ul>
</div></blockquote>
</div></blockquote>
<p>For “ordered” (“ordered_replicon” or “gembase”) datasets only,
MacSyLib can take into account the <strong>shape of the genome</strong>: “linear”,
or “circular” for detection. The default is set to “circular”.</p>
<blockquote>
<div><p>This can be set with the <cite>replicon_topology</cite> parameter from <a class="reference internal" href="#input-options"><span class="std std-ref">Input options</span></a>
or in the configuration in the <a class="reference internal" href="#config-base-label"><span class="std std-ref">base section</span></a>.</p>
<p>With the “gembase” format, it is possible to specify a topology per replicon with a topology file
(see <a class="reference internal" href="gembase_convention.html#gembase-convention"><span class="std std-ref">Gembase format</span></a> and <a class="reference internal" href="gembase_convention.html#topology-files"><span class="std std-ref">Topology files</span></a>).</p>
</div></blockquote>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>MSL can also read <em>.gz</em> compressed files; it will uncompress them on the fly.
The compressed files must end with the <em>.gz</em> extension.
For the <cite>hmmsearch</cite> step You need to have <cite>gunzip</cite> installed on your system for this to work.</p>
</div>
</section>
<section id="config-object">
<span id="config-object-parameters"></span><h2>Config object<a class="headerlink" href="#config-object" title="Link to this heading"></a></h2>
<p>Config object take 2 parameters:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>some default values stored in MacSyDefault (these to class are located in config module)</p></li>
<li><p>and a set of parameters store in an <code class="xref py py-class docutils literal notranslate"><span class="pre">argparse.Namespace</span></code> object</p></li>
</ol>
</div></blockquote>
<p>for instance to set up macsylib</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">argparse</span><span class="w"> </span><span class="kn">import</span> <span class="n">Namespace</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">macsylib.condig</span>

<span class="n">defaults</span> <span class="o">=</span> <span class="n">macsylib</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">MacsyDefaults</span><span class="p">()</span>
<span class="n">settings</span> <span class="o">=</span> <span class="n">Namespace</span><span class="p">(</span>
    <span class="n">db_type</span><span class="o">=</span><span class="s1">&#39;ordered_replicon&#39;</span><span class="p">,</span>
    <span class="n">sequence_db</span><span class="o">=</span><span class="s1">&#39;test.fasta&#39;</span><span class="p">,</span>
    <span class="n">models</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;TXSScan&#39;</span><span class="p">,</span> <span class="s1">&#39;all&#39;</span><span class="p">],</span>  <span class="c1"># this model must be installed with msl_data scripts</span>
    <span class="n">worker</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
<p>The config options available are:</p>
<section id="input-options">
<span id="id1"></span><h3>Input options<a class="headerlink" href="#input-options" title="Link to this heading"></a></h3>
<dl>
<dt>models</dt><dd><p>The models to search. A list, where the first element must be the name of family models, followed by the name of the models.
If the name ‘all’ is in the list all models from the family will be searched.’</p>
<ul class="simple">
<li><p><code class="code docutils literal notranslate"><span class="pre">['TXSScan',</span> <span class="pre">'all']</span></code> to search all models in TXSScan.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">['TXSScan/bacteria',</span> <span class="pre">'all']</span></code> or <code class="code docutils literal notranslate"><span class="pre">['TXSScan',</span> <span class="pre">'bacteria/all']</span></code> to search all models in bacteria.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">['TXSScan/bacteria/diderm',</span> <span class="pre">'T4aP',</span> <span class="pre">'T4bP']</span></code> to search only ‘T4ap’ and ‘T4bP’ models.</p></li>
</ul>
</dd>
<dt>sequence_db</dt><dd><p>Path to the sequence dataset in fasta format.</p>
</dd>
<dt>db_type</dt><dd><p>The type of dataset to deal with.</p>
<blockquote>
<div><ul class="simple">
<li><p><cite>“unordered”*</cite> corresponds to a non-assembled genome,</p></li>
<li><p><cite>“ordered_replicon”</cite> to an assembled genome, where sequence appear in the file as the same order as in the genome.
(be careful the <em>GCF</em> files from the <em>NCBI</em>, most of the time, do not follow the order of the associated <em>gff</em></p></li>
<li><p>and <cite>“gembase”</cite> to a set of replicons see <a class="reference internal" href="gembase_convention.html#gembase-convention"><span class="std std-ref">Gembase format</span></a>.</p></li>
</ul>
</div></blockquote>
</dd>
<dt>replicon_topology</dt><dd><p><cite>circular</cite> or <cite>linear</cite>
The topology of the replicons (this option is meaningful only if the db_type is ‘ordered_replicon’ or ‘gembase’)</p>
</dd>
<dt>topology_file</dt><dd><p>Topology file path. The topology file allows to specify a topology (linear or circular)
for each replicon (this option is meaningful only if the db_type is ‘ordered_replicon’ or ‘gembase’.
A topology file is a tabular file with two columns:</p>
<p>the 1st is the replicon name, and the 2nd the corresponding topology:</p>
<p><code class="code docutils literal notranslate"><span class="pre">RepliconA</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">linear</span></code></p>
</dd>
<dt>idx</dt><dd><p><cite>True</cite> / <cite>False</cite>
Forces to build the indexes for the sequence dataset even if they were previously computed and present at the dataset location.
(default: False)</p>
</dd>
</dl>
</section>
<section id="systems-detection-options">
<span id="system-detect-options"></span><h3>Systems detection options<a class="headerlink" href="#systems-detection-options" title="Link to this heading"></a></h3>
<dl>
<dt>inter_gene_max_space</dt><dd><p>Co-localization criterion: maximum number of components non-matched by a profile allowed between two matched components
for them to be considered contiguous.
Option only meaningful for ‘ordered’ datasets.
This option must be provide as list of tuple.
The tuple first value must match to a model, the second to a number of components.</p>
<p><code class="code docutils literal notranslate"><span class="pre">inter_gene_max_space=[('TXSScan/bacteria/diderm/T2SS',</span> <span class="pre">12),</span> <span class="pre">('TXSScan/bacteria/diderm/Flagellum',</span> <span class="pre">14)],</span></code></p>
</dd>
<dt>min_mandatory_genes_required</dt><dd><p>The minimal number of mandatory genes required for model assessment.
This option must be provide as list of tuple.
The first value must correspond to a model fully qualified name, the second value to an integer.</p>
</dd>
<dt>min_genes_required</dt><dd><p>The minimal number of genes required for model assessment
(includes both ‘mandatory’ and ‘accessory’ components).
The first value must correspond to a model fully qualified name, the second value to an integer.</p>
</dd>
</dl>
<p><code class="code docutils literal notranslate"><span class="pre">min_genes_required=[('TXSScan/bacteria/diderm/T2SS',</span> <span class="pre">5),</span> <span class="pre">('TXSScan/bacteria/diderm/Flagellum',</span> <span class="pre">4)],</span></code></p>
<dl class="simple">
<dt>max-nb-genes</dt><dd><p>The maximal number of genes to consider a system as full.
The first value must correspond to a model name, the second value to an integer.</p>
</dd>
<dt>multi-loci</dt><dd><p>Specifies if the system can be detected as a ‘scattered’ system.
The models are specified as a comma separated list of fully qualified name
<code class="code docutils literal notranslate"><span class="pre">['TXSScan/bacteria/diderm/T2SS',</span> <span class="pre">''TXSScan/bacteria/diderm/Flagellum'']</span></code></p>
</dd>
</dl>
</section>
<section id="options-for-hmmer-execution-and-hits-filtering">
<span id="hmmer-options"></span><h3>Options for Hmmer execution and hits filtering:<a class="headerlink" href="#options-for-hmmer-execution-and-hits-filtering" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>hmmer</dt><dd><p>Path to the hmmsearch program. If it is not specify rely on the PATH. (default: hmmsearch)</p>
</dd>
<dt>e_value_search</dt><dd><p>Maximal e-value for hits to be reported during hmmsearch search.
By default MF set per profile threshold for hmmsearch run (cut_ga option)
for profiles containing the GA bit score threshold.
If a profile does not contains the GA bit score the e_value_search (-E in hmmsearch)
is applied to this profile.
To applied the e_value_search to all profiles use the no_cut_ga option.
(default: 0.1)</p>
</dd>
<dt>no_cut_ga</dt><dd><p>By default the MSL try to applied a threshold per profile by using the hmmer -cut-ga option.
This is possible only if the GA bit score is present in the profile otherwise MSL switch to use the
e_value_search (-E in hmmsearch).
If this option is set the e_value_search option is used for all profiles regardless the presence of
the a GA bit score in the profiles.
(default: False)</p>
</dd>
<dt>cut_ga</dt><dd><p>By default the MSL try to applied a threshold per profile by using the hmmer cut_ga option.
This is possible only if the GA bit score is present in the profile otherwise
MSF switch to use the –e-value-search (-E in hmmsearch).
But the modeler can override this default behavior to do not use cut_ga but e_value_search instead (-E in hmmsearch).
The user can reestablish the general MSL behavior, be sure the profiles contain the GA bit score.
(default: True)</p>
</dd>
<dt>i_evalue_sel</dt><dd><p>Maximal independent e-value for Hmmer hits to be selected for system detection. (default:0.001)</p>
</dd>
<dt>overage_profile</dt><dd><p>Minimal profile coverage required in the hit alignment to allow the hit selection for system detection. (default: 0.5)</p>
</dd>
</dl>
</section>
<section id="options-for-clusters-and-systems-scoring">
<span id="score-options"></span><h3>Options for clusters and systems’ scoring:<a class="headerlink" href="#options-for-clusters-and-systems-scoring" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>mandatory_weight</dt><dd><p>the weight (score) of a mandatory component when scoring clusters (default:1.0)</p>
</dd>
<dt>accessory_weight</dt><dd><p>the weight (score) of an accessory component when scoring clusters (default:0.5)</p>
</dd>
<dt>exchangeable_weight</dt><dd><p>the weight modifier for the score of a component that is exchangeable (default:0.8)</p>
</dd>
<dt>redundancy_penalty</dt><dd><p>the weight modifier for the score of a component that is already present in another cluster</p>
</dd>
<dt>loner_multi_system_weight</dt><dd><p>the weight modifier for the score of a component that is <cite>loner</cite> and <cite>multi-system</cite> at the same time (default:0.7)</p>
</dd>
</dl>
</section>
<section id="path-options">
<span id="id2"></span><h3>Path options:<a class="headerlink" href="#path-options" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>models_dir</dt><dd><p>specify the path to the models if the models are not installed in the canonical place.
It gather definitions (xml files) and hmm profiles in a specific
structure. A directory with the name of the model with at least two directories
profiles” which contains all hmm profile for gene describe in definitions and
models” which contains either xml file of definitions or subdirectories
to organize the model in subsystems.</p>
</dd>
<dt>out_dir</dt><dd><p>Path to the directory where to store results. if out-dir is specified res-search-dir will be ignored.
You have to create it.</p>
</dd>
<dt>force</dt><dd><p>force to run even the out dir already exists and is not empty.
Use this option with caution, MSF will erase everything in out dir before to run. (default= False)</p>
</dd>
<dt>index_dir</dt><dd><p>Specifies the path to a directory to store/read the sequence index when the sequence_db dir is not writable.</p>
</dd>
<dt>res_search_suffix</dt><dd><p>The suffix to give to Hmmer raw output files. (default: .search_hmm.out)</p>
</dd>
<dt>res_extract_suffix</dt><dd><p>The suffix to give to filtered hits output files. (default: .res_hmm_extract)</p>
</dd>
<dt>profile_suffix</dt><dd><p>The suffix of profile files. For each ‘Gene’ element, the corresponding profile is searched in the ‘profile_dir’,
in a file which name is based on the Gene name + the profile suffix.
For instance, if the Gene is named ‘gspG’ and the suffix is ‘.hmm3’, then the profile should be placed at the specified location
and be named ‘gspG.hmm3’ (default: .hmm)</p>
</dd>
</dl>
</section>
<section id="general-options">
<span id="id3"></span><h3>General options:<a class="headerlink" href="#general-options" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>worker</dt><dd><p>Number of workers to be used by MacSylib. In the case the user wants to run MacSyFLib in a multi-thread mode.
(0 mean all threads available will be used (compliant with use of cluster scheduler)). (default: 1)</p>
</dd>
<dt>verbosity</dt><dd><p>Increases the verbosity level. T
here are 4 levels: Error messages (default), Warning (-v), Info (-vv) and Debug.(-vvv)</p>
</dd>
<dt>mute</dt><dd><p>mute the log on stdout.
(continue to log on macsylib.log)
(default: False)</p>
</dd>
<dt>cfg_file</dt><dd><p>Path to a MacSyLib configuration file to be used.</p>
</dd>
<dt>previous_run</dt><dd><p>Path to a previous MacSyLib run directory.
It allows to skip the Hmmer search step on same dataset,
as it uses previous run results and thus parameters regarding Hmmer detection.
The configuration file from this previous run will be used.
Conflict with options: <cite>config</cite>, <cite>sequence_db</cite>, <cite>profile_suffix</cite>, <cite>res_extract_suffix</cite>, <cite>e_value_res</cite>, <cite>db_type</cite>, <cite>hmmer</cite></p>
</dd>
<dt>timeout</dt><dd><p>In some case msf can take a long time to find the best solution (in ‘gembase’ and ‘ordered_replicon mode’).
The timeout is per replicon. If this step reach the timeout, the replicon is skipped (for gembase mode the analyse of other replicons continue).
NUMBER[SUFFIX]  NUMBER seconds. SUFFIX may be ‘s’ for seconds (the default), ‘m’ for minutes, ‘h’ for hours or ‘d’ for days
for instance 1h2m3s means 1 hour 2 min 3 sec. NUMBER must be an integer.</p>
</dd>
</dl>
</section>
</section>
<section id="configuration-file">
<span id="config-definition-label"></span><h2>Configuration file<a class="headerlink" href="#configuration-file" title="Link to this heading"></a></h2>
<p>Options to run MacSyLib can be specified in a configuration file.</p>
<p>The <a class="reference internal" href="../developer_guide/api/config.html#configuration"><span class="std std-ref">Config object</span></a> handles all configuration options for MacSylib.
There kind of locations where to put configuration file:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>System wide configuration (this configuration is used for all macsylib run)</p>
<ul class="simple">
<li><p><em>/etc/macsylib/macsylib.conf</em></p></li>
<li><p>or in <em>${VIRTUAL_ENV}/etc/macsylib.conf</em> if you installed macsylib in a virtualenv</p></li>
<li><p>the file pointed by environment variable <em>MACSY_HOME</em></p></li>
</ul>
</li>
<li><p>User wide configuration (this configuration is used for all run for a user)</p>
<ul class="simple">
<li><p><em>~/.macsylib/macsylib.conf</em></p></li>
</ul>
</li>
<li><p>Project configuration</p>
<ul class="simple">
<li><p><em>macsylib.conf</em> in the current directory</p></li>
<li><p>with command line option <em>–cfg-file</em></p></li>
</ul>
</li>
</ol>
</div></blockquote>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The precedence rules from the least to the most important priority are:</p>
<p>System wide configuration &lt; user wide configuration &lt; project configuration &lt; command line option</p>
</div>
<p>This means that command-line options will always bypass those from the configuration files. In the same flavor,
options altering the definition of systems found in the command-line or the configuration file will always
overwhelm values from systems’ <a class="reference internal" href="../modeler_guide/modeling.html#model-definition-grammar-label"><span class="std std-ref">XML definition files</span></a>.</p>
<p>The configuration files must follow the Python “ini” file syntax.
The <a class="reference internal" href="../developer_guide/api/config.html#configuration"><span class="std std-ref">Config object</span></a> provides some default values and performs some validations of the values.</p>
<p>In MacSyLib, six sections are defined and stored by default in the configuration file:</p>
<blockquote>
<div><blockquote id="config-base-label">
<div><ul>
<li><p><strong>base</strong> : all information related to the protein dataset under study</p>
<ul>
<li><p><em>sequence_db</em> : the path to the dataset in Fasta format (<em>no default value</em>)</p></li>
<li><p><em>db_type</em> : the type of dataset to handle, four types are supported:</p>
<blockquote>
<div><ul class="simple">
<li><p><em>unordered</em> : a set of sequences corresponding to a complete replicon
(<em>e.g.</em> an unassembled complete genome)</p></li>
<li><p><em>ordered_replicon</em> : a set of sequences corresponding to a complete replicon ordered
(<em>e.g.</em> an assembled complete genome)</p></li>
<li><p><em>gembase</em> : a set of multiple ordered replicons.</p></li>
</ul>
</div></blockquote>
<p>(<em>no default value</em>)</p>
</li>
<li><p><em>replicon_topology</em> : the topology of the replicon under study.
Two topologies are supported: ‘linear’ and ‘circular’ (<em>default</em> = ‘circular’).
This option will be ignored if the dataset type is not ordered (<em>i.e.</em> “unordered_replicon” or “unordered”).</p></li>
</ul>
</li>
<li><p><strong>models</strong>
* list of models to search in replicon</p></li>
<li><p><strong>models_opt</strong></p>
<ul class="simple">
<li><p><em>inter_gene_max_space</em> = list of models’ fully qualified names and integer separated by spaces (see example below).
These values will supersede the values found in the model definition file.</p></li>
<li><p><em>min_mandatory_genes_required</em> = list of models’ fully qualified name and integer separated by spaces.
These values will supersede the values found in the model definition file.</p></li>
<li><p><em>min_genes_required</em> = list of models’ fully qualified name and integer separated by spaces.
These values will supersede the values found in the model definition file.</p></li>
<li><p><em>max_nb_genes</em> = list of models’ fully qualified names and integer separated by spaces.
These values will supersede the values found in the model definition file.</p></li>
</ul>
</li>
<li><p><strong>hmmer</strong></p>
<ul class="simple">
<li><p><em>hmmer_exe</em> (default= <em>hmmsearch</em> )</p></li>
<li><p><em>e_value_res</em> = (default= <em>1</em> )</p></li>
<li><p><em>i_evalue_sel</em> = (default= <em>0.5</em> )</p></li>
<li><p><em>coverage_profile</em> = (default= <em>0.5</em> )</p></li>
</ul>
</li>
<li><p><strong>score_opt</strong></p>
<ul class="simple">
<li><p><em>mandatory_weight</em> (default= <em>1.0</em>)</p></li>
<li><p><em>accessory_weight</em> (default= <em>0.5</em>)</p></li>
<li><p><em>exchangeable_weight</em> (default= <em>0.8</em>)</p></li>
<li><p><em>redundancy_penalty</em> (default= <em>1.5</em>)</p></li>
<li><p><em>out_of_cluster</em> (default= <em>0.7</em>)</p></li>
</ul>
</li>
<li><p><strong>directories</strong></p>
<ul class="simple">
<li><p><em>res_search_dir</em> = (default= <em>./datatest/res_search</em> )</p></li>
<li><p><em>res_search_suffix</em> = (default= <em>.search_hmm.out</em> )</p></li>
<li><p><em>system_models_dir</em> = (default= <em>./models</em> )</p></li>
<li><p><em>res_extract_suffix</em> = (default= <em>.res_hmm_extract</em> )</p></li>
<li><p><em>index_dir</em> = (default= beside the sequence_db)</p></li>
</ul>
</li>
<li><p><strong>general</strong></p>
<ul>
<li><dl>
<dt><em>log_level</em>: (default= <em>debug</em> ) This corresponds to an integer code:</dt><dd><table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Level</p></th>
<th class="head"><p>Numeric value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>CRITICAL</p></td>
<td><p>50</p></td>
</tr>
<tr class="row-odd"><td><p>ERROR</p></td>
<td><p>40</p></td>
</tr>
<tr class="row-even"><td><p>WARNING</p></td>
<td><p>30</p></td>
</tr>
<tr class="row-odd"><td><p>INFO</p></td>
<td><p>20</p></td>
</tr>
<tr class="row-even"><td><p>DEBUG</p></td>
<td><p>10</p></td>
</tr>
<tr class="row-odd"><td><p>NOTSET</p></td>
<td><p>0</p></td>
</tr>
</tbody>
</table>
</dd>
</dl>
</li>
<li><p><em>log_file</em> = (default = macsylib.log in directory of the run)</p></li>
</ul>
</li>
</ul>
</div></blockquote>
</div></blockquote>
<p>Example of a configuration file</p>
<div class="highlight-ini notranslate"><div class="highlight"><pre><span></span><span class="w"> </span><span class="k">[base]</span>
<span class="w"> </span><span class="na">prefix</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">/path/to/macsylib/home/</span>
<span class="w"> </span><span class="na">file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">%(prefix)s/data/base/prru_psae.001.c01.fasta</span>
<span class="w"> </span><span class="na">db_type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">gembase</span>
<span class="w"> </span><span class="na">replicon_topology</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">circular</span>

<span class="w"> </span><span class="k">[models]</span>
<span class="w"> </span><span class="na">models_1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">TFF-SF_final all</span>

<span class="w"> </span><span class="k">[models_opt]</span>
<span class="w"> </span><span class="na">inter_gene_max_space</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">TXSS/T2SS 22 TXSS/Flagellum 44</span>
<span class="w"> </span><span class="na">min_mandatory_genes_required</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">TXSS/T2SS 6 TXSS/Flagellum 4</span>
<span class="w"> </span><span class="na">min_genes_required</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">TXSS/T2SS 8 TXSS/Flagellum 4</span>
<span class="w"> </span><span class="na">max_nb_genes</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">TXSS/T2SS 12 TXSS/Flagellum 8</span>

<span class="w"> </span><span class="k">[hmmer]</span>
<span class="w"> </span><span class="na">hmmer</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">hmmsearch</span>
<span class="w"> </span><span class="na">e_value_res</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">1</span>
<span class="w"> </span><span class="na">i_evalue_sel</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">0.5</span>
<span class="w"> </span><span class="na">coverage_profile</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">0.5</span>

<span class="w"> </span><span class="k">[score_opt]</span>
<span class="w"> </span><span class="na">mandatory_weight</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">1.0</span>
<span class="w"> </span><span class="na">accessory_weight</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">0.5</span>
<span class="w"> </span><span class="na">exchangeable_weight</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">0.8</span>
<span class="w"> </span><span class="na">redundancy_penalty</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">1.5</span>
<span class="w"> </span><span class="na">loner_multi_system_weight</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">0.7</span>

<span class="w"> </span><span class="k">[directories]</span>
<span class="w"> </span><span class="na">prefix</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">/path/to/macsylib/home/</span>
<span class="w"> </span><span class="na">data_dir</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">%(prefix)s/data/</span>
<span class="w"> </span><span class="na">res_search_dir</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">%(prefix)s/dataset/res_search/</span>
<span class="w"> </span><span class="na">res_search_suffix</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">.raw_hmm</span>
<span class="w"> </span><span class="na">system_models_dir</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">%(data_dir)/data/models, ~/.macsylib/data</span>
<span class="w"> </span><span class="na">profile_suffix</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">.fasta-aln.hmm</span>
<span class="w"> </span><span class="na">res_extract_suffix</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">.res_hmm</span>
<span class="w"> </span><span class="na">index_dir</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">path/where/I/store/my_indexes</span>

<span class="k">[general]</span>
<span class="na">log_level</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">debug</span>
<span class="na">worker</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">4</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>After a run, the corresponding configuration file (“macsylib.conf”) is generated as a (re-usable)
output file that stores every options used in the run.
It is stored in the results’ directory (see <a class="reference internal" href="outputs.html#outputs"><span class="std std-ref">the output section</span></a>).</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The configuration variable <cite>models_dir</cite> cannot be set in general configuration file.
<cite>models_dir`</cite> can be set only in configuration under user control.
<code class="docutils literal notranslate"><span class="pre">`$(HOME)/.macsylib/macsylib.conf</span> <span class="pre">&lt;</span> <span class="pre">macsylib.conf</span> <span class="pre">&lt;</span> <span class="pre">&quot;command-line&quot;</span> <span class="pre">options`</span></code>
<cite>models_dir</cite> is a single path to a directory whre masyfinder can find models.</p>
<p>But the <cite>system_models_dir</cite> can be set in general configuration file</p>
<ul class="simple">
<li><p>/etc/macsylib/macsylib.conf</p></li>
<li><p>or ${VIRTUAL_ENV}/etc/macsylib/macsylib.conf</p></li>
<li><p>or anywhere point by $MACSY_CONF environment variable</p></li>
</ul>
<p><cite>system_models_dir</cite> manage a list of locations where macsylib can find models.
The order of locations is important, it reflects the precedence rule (The models found in last location
superseed models found in previous location).
By default look for following directories: <em>/share/macsylib/models</em>, or <em>/usr/share/macsylib/models</em>
and <em>$HOME/.macsylib/models</em> and <cite>system_models_dir</cite> uses these directories if they exists.</p>
</div>
</section>
<section id="in-house-input-files">
<h2>In-house input files<a class="headerlink" href="#in-house-input-files" title="Link to this heading"></a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="gembase_convention.html">Gembase format</a></li>
<li class="toctree-l1"><a class="reference internal" href="gembase_convention.html#topology-files">Topology files</a></li>
</ul>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="quickstart.html" class="btn btn-neutral float-left" title="MacSyLib Quick Start" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="gembase_convention.html" class="btn btn-neutral float-right" title="Gembase format" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2014-2025, Bertrand Néron, Sophie Abby.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
  

<div class="licensing">
  <p>
    This documentation is licensed under <a href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License
      <img src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png">.
    </a>
  </p>
</div>


</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

<a class="github" href="https://github.com/gem-pasteur/macsylib">
    <img style="position: absolute; top: 0; right: 0; border: 0;"
         src="https://raw.githubusercontent.com/gem-pasteur/macsylib/refs/heads/main/doc/source/_static/github-banner.svg"
         alt="Fork me on GitHub" />
</a>



</body>
</html>