

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>MacSyLib implementation overview &mdash; MacSyLib 1.0.3 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/graphviz.css?v=4ae1632d" />
      <link rel="stylesheet" type="text/css" href="../_static/_css/custom.css?v=b89f2dfc" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=baaebd52"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="MacSyLib API documentation" href="api/index.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            MacSyLib
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../user_guide/index.html">User Guide</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modeler_guide/index.html">Modeller Guide</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Developer Guide</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">MacSyLib implementation overview</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#macsylib-project-structure">MacSyLib project structure</a></li>
<li class="toctree-l3"><a class="reference internal" href="#macsylib-architecture-overview">MacSyLib architecture overview</a></li>
<li class="toctree-l3"><a class="reference internal" href="#search-system-functioning-overview">search_system functioning overview</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#the-model-object">The Model object</a></li>
<li class="toctree-l4"><a class="reference internal" href="#the-gene-object">The Gene object</a></li>
<li class="toctree-l4"><a class="reference internal" href="#the-profile-object">The Profile object</a></li>
<li class="toctree-l4"><a class="reference internal" href="#reporting-hmmer-search-results">Reporting Hmmer search results</a></li>
</ul>
</li>
<li><a class="reference internal" href="#tests-coverage">tests coverage</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="api/index.html">MacSyLib API documentation</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">MacSyLib</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">Developer Guide</a></li>
      <li class="breadcrumb-item active">MacSyLib implementation overview</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="macsylib-implementation-overview">
<span id="overview"></span><h1>MacSyLib implementation overview<a class="headerlink" href="#macsylib-implementation-overview" title="Link to this heading"></a></h1>
<p>MacSyLib is implemented with an object-oriented architecture.
Below a short glossary to fix the vocabulary used in MacSylib.</p>
<dl class="simple glossary">
<dt id="term-Cluster">Cluster<a class="headerlink" href="#term-Cluster" title="Link to this term"></a></dt><dd><p>Is a “contiguous” set of hits.
two hits are considered contiguous if the number of genes between the 2 genes matching the 2 hits
in the replicon is lesser than inter-genes-max-space.</p>
</dd>
<dt id="term-Model">Model<a class="headerlink" href="#term-Model" title="Link to this term"></a></dt><dd><p>Is a formal description of a macromolecular system.
Is composed of a definition and a list of profiles.
at each gene of the Model must correspond a profile</p>
</dd>
<dt id="term-Model-family">Model family<a class="headerlink" href="#term-Model-family" title="Link to this term"></a></dt><dd><p>A set of models, on the same topic.
It is composed of several definitions which can be sorted in hierachical structure
and profiles. A profile is a hmm profile file.</p>
</dd>
<dt id="term-ModelDefinition">ModelDefinition<a class="headerlink" href="#term-ModelDefinition" title="Link to this term"></a></dt><dd><p>Is a definition of model, it’s serialize as a xml file</p>
</dd>
<dt id="term-Solution">Solution<a class="headerlink" href="#term-Solution" title="Link to this term"></a></dt><dd><p>It’s a systems combination for one replicon.
The best solution for a replicon, is the combination of all systems found in this replicon which
maximize the score.</p>
</dd>
<dt id="term-System">System<a class="headerlink" href="#term-System" title="Link to this term"></a></dt><dd><p>It’s an occurrence of a specific Model on a replicon.
Basically, it’s a cluster or set of clusters which satisfy the Model quorum.</p>
</dd>
</dl>
<section id="macsylib-project-structure">
<h2>MacSyLib project structure<a class="headerlink" href="#macsylib-project-structure" title="Link to this heading"></a></h2>
<p>A brief overview of the files and directory constituting the MacSylib project.</p>
<dl class="simple glossary">
<dt id="term-doc">doc<a class="headerlink" href="#term-doc" title="Link to this term"></a></dt><dd><p>The project is documented using sphinx.
All sources files needed to generate this documentation is in the directory <em>doc</em></p>
</dd>
<dt id="term-macsylib">macsylib<a class="headerlink" href="#term-macsylib" title="Link to this term"></a></dt><dd><p>This the MacSyLib python library
Inside macsypy there is a subdirectory <em>scripts</em> which are the entry points for
<cite>macsyprofile</cite> (tool to help modeler to analyse <em>hmmsearch</em> output) and <cite>macsydata</cite> (tool to install models)</p>
</dd>
<dt id="term-tests">tests<a class="headerlink" href="#term-tests" title="Link to this term"></a></dt><dd><p>The code is tested using <cite>unittests</cite>.
In <em>tests</em> the directory <em>data</em> contains all data needed to perform the tests.</p>
</dd>
<dt id="term-CONTRIBUTORS">CONTRIBUTORS<a class="headerlink" href="#term-CONTRIBUTORS" title="Link to this term"></a></dt><dd><p>A file containing the list of code contributors.</p>
</dd>
<dt id="term-CONTRIBUTING">CONTRIBUTING<a class="headerlink" href="#term-CONTRIBUTING" title="Link to this term"></a></dt><dd><p>A guide on how to contribute to the project.</p>
</dd>
<dt id="term-COPYRIGHT">COPYRIGHT<a class="headerlink" href="#term-COPYRIGHT" title="Link to this term"></a></dt><dd><p>The MacSyLib copyrights.</p>
</dd>
<dt id="term-COPYING">COPYING<a class="headerlink" href="#term-COPYING" title="Link to this term"></a></dt><dd><p>The licencing.
MacSyLib is released under GPLv3.</p>
</dd>
<dt id="term-README.md">README.md<a class="headerlink" href="#term-README.md" title="Link to this term"></a></dt><dd><p>Brief information about the project.</p>
</dd>
<dt id="term-setup.py">setup.py<a class="headerlink" href="#term-setup.py" title="Link to this term"></a></dt><dd><p>The installation recipe.</p>
</dd>
<dt id="term-pyproject.toml">pyproject.toml<a class="headerlink" href="#term-pyproject.toml" title="Link to this term"></a></dt><dd><p>The project information and installation recipe.</p>
</dd>
<dt id="term-codemeta.json">codemeta.json<a class="headerlink" href="#term-codemeta.json" title="Link to this term"></a></dt><dd><p>metadata on this project. (<a class="reference external" href="https://codemeta.github.io/">https://codemeta.github.io/</a>)</p>
</dd>
</dl>
</section>
<section id="macsylib-architecture-overview">
<h2>MacSyLib architecture overview<a class="headerlink" href="#macsylib-architecture-overview" title="Link to this heading"></a></h2>
<p>An overview of the main classes.</p>
<figure class="align-default" id="id1">
<img alt="../_images/macsylib_classes.drawio.svg" src="../_images/macsylib_classes.drawio.svg" />
<figcaption>
<p><span class="caption-text">The MacSyLib classes diagram.
The classes are not details. only the main attributes allowing us to understand the interaction are mentioned.</span><a class="headerlink" href="#id1" title="Link to this image"></a></p>
<div class="legend">
<ul class="simple">
<li><p>in green the modules</p></li>
<li><p>in orange, the concrete class</p></li>
<li><p>in red the abstract classes</p></li>
<li><p>in blue the enumeration</p></li>
<li><p>in purple the dataclass</p></li>
<li><p>in purple/pink functions</p></li>
</ul>
</div>
</figcaption>
</figure>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>use <em>view image</em> of your browser to  zoom in the diagram</p>
</div>
</section>
<section id="search-system-functioning-overview">
<h2>search_system functioning overview<a class="headerlink" href="#search-system-functioning-overview" title="Link to this heading"></a></h2>
<p>In this section I’ll give you an idea of the <code class="xref py py-func docutils literal notranslate"><span class="pre">macsylib.search_systems.search_systems()</span></code> functioning
at very high grain coarse.</p>
<p>The higher level function is <code class="xref py py-func docutils literal notranslate"><span class="pre">macsylib.search_systems.search_systems()</span></code>.
But to call this function you have to create a <a class="reference internal" href="api/config.html#macsylib.config.Config" title="macsylib.config.Config"><code class="xref py py-class docutils literal notranslate"><span class="pre">macsylib.config.Config</span></code></a> object (<a class="reference internal" href="api/config.html#configuration"><span class="std std-ref">configuration</span></a>)
and also initialize the logger.</p>
<p>The first <cite>search_systems</cite> task is to create models asked by the user.
So a <a class="reference internal" href="api/definition_parser.html#macsylib.definition_parser.DefinitionParser" title="macsylib.definition_parser.DefinitionParser"><code class="xref py py-class docutils literal notranslate"><span class="pre">macsylib.definition_parser.DefinitionParser</span></code></a> is instantiated and the <a class="reference internal" href="api/model.html#macsylib.model.ModelBank" title="macsylib.model.ModelBank"><code class="xref py py-class docutils literal notranslate"><span class="pre">macsylib.model.ModelBank</span></code></a>
and <a class="reference internal" href="api/gene.html#macsylib.gene.GeneBank" title="macsylib.gene.GeneBank"><code class="xref py py-class docutils literal notranslate"><span class="pre">macsylib.gene.GeneBank</span></code></a> are populated.</p>
<p>Once all models are created, we gather all genes and search them in the replicons.
This step is done in parallel.
The search is done by profile object associated to each gene and rely on the external software <em>hmmsearch</em>.
The parallelization is ensure by search_genes function
The results of this step is a list of hits.</p>
<p>This list is sorted by position and score.
this list is filtered to keep only one hit for each position,
the one with the best score (position is a gene product in a replicon)</p>
<p>For each model asked by the user, we filter the hits list to keep only those related to the model.
Those which are link to <cite>mandatory</cite>, <cite>accessory</cite>, <cite>neutral</cite> or <cite>forbidden</cite> genes included the exchangeables.</p>
<p>This hits are clustered based on distance constraints describe in the models:</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>inter_gene_max_space</strong> : the maximum genes allowed between to genes of a system.</p></li>
<li><p><strong>loner</strong> : allow a gene to participate to system even if it does not clusterize with some other genes.</p></li>
</ul>
</div></blockquote>
<p>Then we check if each cluster satisfy the quorum described in the model.</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>min_mandatory_genes</strong> : the minimum of mandatory genes requisite to have a system.</p></li>
<li><p><strong>min_genes_required</strong> : the minimum of genes (mandatory + accessory) requisite to have a system.</p></li>
<li><p><strong>forbidden_genes</strong> : no forbidden genes may appear in the cluster.</p></li>
</ul>
</div></blockquote>
<p>If the model is <em>multi_loci</em> we generate a combination of the clusters and check the quorum for each combination.
If the cluster or combination satisfy the quorum a <code class="xref py py-class docutils literal notranslate"><span class="pre">macsypy.systems.System</span></code> is created otherwise a
<code class="xref py py-class docutils literal notranslate"><span class="pre">macsypy.cluster.RejectedCandidate</span></code>.</p>
<p>The Systems from the same replicon are sort against their position, score.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <em>neutral</em> genes are used to build clusters. But not to fulfill the quorum.</p>
</div>
<p>Among all this potential systems, MSL (MacSyLib’ compute the best combination. <a class="reference internal" href="api/solution.html#macsylib.solution.find_best_solutions" title="macsylib.solution.find_best_solutions"><code class="xref py py-func docutils literal notranslate"><span class="pre">macsylib.solution.find_best_solutions()</span></code></a>.
The best combination is the set of compatible systems (do not share common hits) which maximize the score.
It’s possible to have several equivalent “best solutions”.
The results of this step is reported in the <cite>best_systems.tsv</cite> file.</p>
<section id="the-model-object">
<span id="system-implementation"></span><h3>The Model object<a class="headerlink" href="#the-model-object" title="Link to this heading"></a></h3>
<p>The <a class="reference internal" href="api/model.html#model"><span class="std std-ref">Model object</span></a> represents a macromolecular model to detect.
It is defined <em>via</em> a definition file in XML stored in a dedicated location that can be specified <em>via</em>
the <a class="reference internal" href="api/config.html#macsylib.config.Config" title="macsylib.config.Config"><code class="xref py py-class docutils literal notranslate"><span class="pre">macsylib.config.Config</span></code></a> object.
See <a class="reference internal" href="../modeler_guide/modeling.html#model-definition-grammar-label"><span class="std std-ref">The XML hierarchy</span></a> for more details on the XML grammar.</p>
<p>An object <a class="reference internal" href="api/definition_parser.html#definition-parser"><span class="std std-ref">ModelDefinitionParser</span></a> is used to build a model object from its XML definition file.</p>
<p>A model is named after the file tree name of its XML definition.
A model has an attribute <cite>inter_gene_max_space</cite> which is an integer,
and four kind of components are listed in function of their presence in the system:</p>
<ul class="simple">
<li><p>The genes that must be present in the genome to define this model (“mandatory”).</p></li>
<li><p>The genes that can be present, but do not have to be found in every case (“accessory”).</p></li>
<li><p>The genes that are used to build clusters, but not take in account to check the quorum
(<code class="docutils literal notranslate"><span class="pre">min-genes-required</span></code> and <code class="docutils literal notranslate"><span class="pre">min-mandatory-genes-required</span></code>) are described as “neutral”.</p></li>
<li><p>The genes that must not be present in the system (“forbidden”).</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>A complete description of macromolecular models modelling is available in the section <a class="reference internal" href="../user_guide/model_definition.html#model-definition"><span class="std std-ref">Macromolecular models</span></a></p>
</div>
</section>
<section id="the-gene-object">
<span id="gene-implementation"></span><h3>The Gene object<a class="headerlink" href="#the-gene-object" title="Link to this heading"></a></h3>
<p>The <a class="reference internal" href="api/gene.html#id1"><span class="std std-ref">Gene object</span></a> represents genes encoding the protein components of a Model.
There is 2 kind of gene The <code class="docutils literal notranslate"><span class="pre">CoreGene</span></code> (<a class="reference internal" href="api/gene.html#macsylib.gene.CoreGene" title="macsylib.gene.CoreGene"><code class="xref py py-class docutils literal notranslate"><span class="pre">macsylib.gene.CoreGene</span></code></a>) which must be unique given a name.
A <code class="docutils literal notranslate"><span class="pre">CoreGene</span></code> must have a corresponding HMM protein profile.
These profiles are represented by Profile objects (<a class="reference internal" href="api/profile.html#macsylib.profile.Profile" title="macsylib.profile.Profile"><code class="xref py py-class docutils literal notranslate"><span class="pre">macsylib.profile.Profile</span></code></a>),
and must be named after the gene name. For instance, the gene <em>gspD</em> will correspond to the “gspD.hmm” profile file.
See <a class="reference internal" href="#profile-implementation"><span class="std std-ref">The Profile object</span></a>). After <em>hmmsearch</em> step the hits are link the them.
The <a class="reference internal" href="api/gene.html#macsylib.gene.CoreGene" title="macsylib.gene.CoreGene"><code class="xref py py-class docutils literal notranslate"><span class="pre">macsylib.gene.CoreGene</span></code></a> objects must be created by using the <a class="reference internal" href="api/gene.html#macsylib.gene.GeneBank" title="macsylib.gene.GeneBank"><code class="xref py py-class docutils literal notranslate"><span class="pre">macsylib.gene.GeneBank</span></code></a> factory.</p>
<p>A <code class="docutils literal notranslate"><span class="pre">ModelGene</span></code> (<a class="reference internal" href="api/gene.html#macsylib.gene.ModelGene" title="macsylib.gene.ModelGene"><code class="xref py py-class docutils literal notranslate"><span class="pre">macsylib.gene.ModelGene</span></code></a>) which encapsulate a CoreGene and is linked to a Model.
Instead <code class="docutils literal notranslate"><span class="pre">CoreGene</span></code>, several <code class="docutils literal notranslate"><span class="pre">ModelGene</span></code> with the same name may coexists in one macsylib run,
in different Models and hold different values for attributes as <em>inter_gene_max_space</em>, …
Each ModelGene points out its Model of origin (<a class="reference internal" href="api/model.html#macsylib.model.Model" title="macsylib.model.Model"><code class="xref py py-class docutils literal notranslate"><span class="pre">macsylib.model.Model</span></code></a>).
A Gene has several properties described in the <a class="reference internal" href="api/gene.html#id1"><span class="std std-ref">Gene API</span></a>.</p>
<p>A ModelGene may be functionally replaced by an other (usually Homologs or Analogs).
In this case these genes are described as exchangeables.
Exchangeable object encapsulates a ModelGene and has a reference to the ModelGene it is exchangeable to.
See the <a class="reference internal" href="api/gene.html#exchangeable-api"><span class="std std-ref">Exchangeable API</span></a> for more details.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>To optimize computation and to avoid concurrency problems when we search several Models,
each CoreGene must be instantiated only once, and stored in a <em>“gene_bank”</em>.
gene_bank is a <a class="reference internal" href="api/gene.html#macsylib.gene.GeneBank" title="macsylib.gene.GeneBank"><code class="xref py py-class docutils literal notranslate"><span class="pre">macsylib.gene.GeneBank</span></code></a> object.
The <em>gene_bank</em> and <em>model_bank</em> are filled by the <em>system_parser</em> (<code class="xref py py-class docutils literal notranslate"><span class="pre">macsylib.definition_parser.ModelDefinitionParser</span></code>)</p>
</div>
</section>
<section id="the-profile-object">
<span id="profile-implementation"></span><h3>The Profile object<a class="headerlink" href="#the-profile-object" title="Link to this heading"></a></h3>
<p>Each <em>“CoreGene”</em> component corresponds to a <em>“Profile”</em>.
The <a class="reference internal" href="api/profile.html#macsylib.profile.Profile" title="macsylib.profile.Profile"><code class="xref py py-class docutils literal notranslate"><span class="pre">macsylib.profile.Profile</span></code></a> object is used for the search of the gene with <em>Hmmer</em>.
Thus, a <em>“Profile”</em> must match a HMM file, which name is based on the profile name.
For instance, the <em>gspG</em> gene has the corresponding “gspG.hmm” profile file provided at a dedicated location.</p>
</section>
<section id="reporting-hmmer-search-results">
<span id="report-implementation"></span><h3>Reporting Hmmer search results<a class="headerlink" href="#reporting-hmmer-search-results" title="Link to this heading"></a></h3>
<p>A <em>“HMMReport”</em> (<a class="reference internal" href="api/report.html#macsylib.report.HMMReport" title="macsylib.report.HMMReport"><code class="xref py py-class docutils literal notranslate"><span class="pre">macsylib.report.HMMReport</span></code></a>) object represents the results of a Hmmer program search on
the input dataset with a hidden Markov model protein profile.
This object has methods to extract and build <em>“Hits”</em> that are then analyzed for systems assessment.</p>
<p>It analyses Hmmer raw outputs, and applies filters on the matches (according to <a class="reference internal" href="../user_guide/input.html#hmmer-options"><span class="std std-ref">Hmmer options</span></a>).
See <a class="reference internal" href="../user_guide/outputs.html#hmmer-outputs-label"><span class="std std-ref">Hmmer results’ output files</span></a> for details on the resulting output files.
For profile matches selected with the filtering parameters, <em>“Hit”</em> objects are built (see <a class="reference internal" href="api/hit.html#hit"><span class="std std-ref">the Hit API</span></a>).</p>
</section>
</section>
<section id="tests-coverage">
<h2>tests coverage<a class="headerlink" href="#tests-coverage" title="Link to this heading"></a></h2>
<p><a class="reference external" href="http://gem.pages.pasteur.fr/macsylib/coverage">macsylib coverage</a></p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="api/index.html" class="btn btn-neutral float-right" title="MacSyLib API documentation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2014-2025, Bertrand Néron, Sophie Abby.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
  

<div class="licensing">
  <p>
    This documentation is licensed under <a href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License
      <img src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png">.
    </a>
  </p>
</div>


</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

<a class="github" href="https://github.com/gem-pasteur/macsylib">
    <img style="position: absolute; top: 0; right: 0; border: 0;"
         src="https://raw.githubusercontent.com/gem-pasteur/macsylib/refs/heads/main/doc/source/_static/github-banner.svg"
         alt="Fork me on GitHub" />
</a>



</body>
</html>