

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>MacSyFinder Quick Start &mdash; MacSyFinder 2.1.5 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/graphviz.css?v=4ae1632d" />
      <link rel="stylesheet" type="text/css" href="../_static/_css/custom.css?v=b89f2dfc" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=786ba454"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Input and Options of MacSyFinder" href="input.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            MacSyFinder
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">User Guide</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="index.html#running-macsyfinder">Running MacSyFinder</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="new_v2.html">What’s new in MacSyFinder v2?</a></li>
<li class="toctree-l3"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">MacSyFinder Quick Start</a></li>
<li class="toctree-l3"><a class="reference internal" href="#an-example-data-set">An example data set</a></li>
<li class="toctree-l3"><a class="reference internal" href="input.html">Input and Options of MacSyFinder</a></li>
<li class="toctree-l3"><a class="reference internal" href="outputs.html">Output format</a></li>
<li class="toctree-l3"><a class="reference internal" href="big_data.html">For big data people</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="index.html#macsyfinder-functioning">MacSyFinder functioning</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#frequently-asked-questions">Frequently Asked Questions</a></li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modeler_guide/index.html">Modeller Guide</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide/index.html">Developer Guide</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">MacSyFinder</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">User Guide</a></li>
      <li class="breadcrumb-item active">MacSyFinder Quick Start</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="macsyfinder-quick-start">
<span id="quickstart"></span><h1>MacSyFinder Quick Start<a class="headerlink" href="#macsyfinder-quick-start" title="Link to this heading"></a></h1>
<ol class="arabic">
<li><p>We recommend to install MacSyFinder using <cite>pip</cite> in a virtual environment (for further details see <a class="reference internal" href="installation.html#user-installation"><span class="std std-ref">Installation</span></a>).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python3<span class="w"> </span>-m<span class="w"> </span>venv<span class="w"> </span>MacSyFinder
<span class="nb">cd</span><span class="w"> </span>MacSyFinder
<span class="nb">source</span><span class="w"> </span>bin/activate
pip<span class="w"> </span>install<span class="w"> </span>macsyfinder
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p><cite>hmmsearch</cite> from the HMMER package (<a class="reference external" href="http://hmmer.org/">http://hmmer.org/</a>) must be installed.</p>
</div>
</li>
<li><p>Prepare your data. You need a file containing all protein sequences of your genome of interest in a FASTA file
(for further details see <a class="reference internal" href="input.html#input-dataset-label"><span class="std std-ref">Input dataset</span></a>). In the best case scenario, they would be ordered as the
corresponding genes are ordered along the replicons.</p></li>
<li><p>You need to install, or make available to MacSyFinder the models to search in your input genome data.
Please refer to <a class="reference internal" href="model_definition.html#model-definition"><span class="std std-ref">Macromolecular models</span></a> to create your own package of models.
Otherwise, macsy-models contributed by the community are available here: <a class="reference external" href="https://github.com/macsy-models">https://github.com/macsy-models</a>
and can be retrieved and installed using the <a class="reference internal" href="installation.html#msf-data"><span class="std std-ref">msf_data</span></a> command, installed as part of the MacSyFinder suite.</p></li>
<li><p>Command lines:</p>
<blockquote>
<div><ul>
<li><p>Type:
<code class="code docutils literal notranslate"><span class="pre">macsyfinder</span> <span class="pre">-h</span></code></p>
<p>To see all the options available. All command-line options are described in the <a class="reference internal" href="input.html#command-line-label"><span class="std std-ref">Command-line options section</span></a>.
In order to run MacSyFinder on your favorite dataset as soon as you have installed the macsy-model of interest,
you can simply follow the following steps:</p>
</li>
<li><p>Install the macsy-models of interest from the <a class="reference external" href="https://github.com/macsy-models">Macsy Models repository</a>:</p>
<p><code class="code docutils literal notranslate"><span class="pre">msf_data</span> <span class="pre">install</span> <span class="pre">some-public-models</span></code></p>
</li>
<li><p>On a “unordered” genome dataset:</p>
<p><code class="code docutils literal notranslate"><span class="pre">macsyfinder</span> <span class="pre">--db-type</span> <span class="pre">unordered</span> <span class="pre">--sequence-db</span> <span class="pre">unordered_genome.fasta</span> <span class="pre">--models</span> <span class="pre">model_family</span> <span class="pre">all</span></code></p>
<p>will search for systems corresponding to all the models of <cite>model_family</cite> modeled in .xml files shipped with the <em>“some-public-models”</em>
macsy-model package, without taking into account the gene order.</p>
</li>
<li><p>On a completely assembled genome (where the gene order is known):</p>
<p><code class="code docutils literal notranslate"><span class="pre">macsyfinder</span> <span class="pre">--db-type</span> <span class="pre">ordered_replicon</span> <span class="pre">--sequence-db</span> <span class="pre">mygenome.fasta</span> <span class="pre">--models-dir</span> <span class="pre">my-models</span> <span class="pre">--models</span> <span class="pre">model_family</span> <span class="pre">ModelA</span> <span class="pre">ModelB</span></code></p>
<p>will detect the macromolecular systems described in the two models <em>“ModelA”</em> and <em>“ModelB”</em>
in a complete genome from the <em>“ModelA.xml”</em> and <em>“ModelB.xml”</em>
definition files placed in the folder <em>“my-models/model_family/definitions”</em>.</p>
</li>
<li><p>If you want to run the same analysis as above but with local macsy-models not installed by msf_data:</p>
<p><code class="code docutils literal notranslate"><span class="pre">macsyfinder</span> <span class="pre">--db-type</span> <span class="pre">ordered_replicon</span> <span class="pre">--sequence-db</span> <span class="pre">mygenome.fasta</span> <span class="pre">--models-dir</span> <span class="pre">my-models</span> <span class="pre">--models</span> <span class="pre">model_family</span> <span class="pre">ModelA</span> <span class="pre">ModelB</span></code></p>
<p><cite>my-models</cite> is the directory containing  the macsy-model packages.
NB: The models must follow the <a class="reference internal" href="../modeler_guide/package.html#package-structure"><span class="std std-ref">macsy-models package</span></a> structure.</p>
</li>
</ul>
</div></blockquote>
</li>
</ol>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Systems names have to be spelled in a case-sensitive way to run their detection from the command-line.
The name of the System corresponds to the suffix defined for xml files (.xml by default),
for example <em>“toto”</em> for a model defined in <em>“toto.xml”</em>.</p>
<p>The <em>“all”</em> keyword allows to detect all models available in the definitions folder in a single run.
See the <a class="reference internal" href="input.html#command-line-label"><span class="std std-ref">Command-line options</span></a>.</p>
</div>
</section>
<section id="an-example-data-set">
<h1>An example data set<a class="headerlink" href="#an-example-data-set" title="Link to this heading"></a></h1>
<blockquote>
<div><p>We provide <a class="reference external" href="https://doi.org/10.6084/m9.figshare.21581280.v1">here</a> an example dataset comprising a replicon
and the output files expected with MacSyFinder, release 2.0 when running the TXSScan macsy-models.
The genomic dataset consists in the complete sequence of chromosome I from <cite>Vibrio cholerae</cite> O1 biovar El Tor str. N16961
(published here: <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/10952301/">https://pubmed.ncbi.nlm.nih.gov/10952301/</a>).</p>
<p>The chromosome to annotate is presented as a multi-FASTA file of the proteins ordered as the genes encoding them.
An annotation of the protein secretion systems and appendages was run on the genome, using the macsyfinder set of models (“macsy-model”) TXSScan, V1.1.1 in the case of these examples.
There are two output files offered, the one expected with the “ordered” genome mode of annotation, and the other with the “unordered” mode of genome annotation.
The following command lines were used to obtain the output files:</p>
<p>1. The genome is downloaded from <a class="reference external" href="https://doi.org/10.6084/m9.figshare.21581280.v1">here</a>.
It will serve as an input file in the next command-line examples.</p>
<p>2. The TXSScan models for annotation of secretion systems are installed.
The command line is the following:</p>
<p><code class="code docutils literal notranslate"><span class="pre">msf_data</span> <span class="pre">install</span> <span class="pre">TXSScan</span></code>
<cite># Installs the latest version of TXSScan</cite></p>
<ol class="arabic" start="3">
<li><p>MacSyFinder is run on the genome, here using 8 workers for the HMM search (“-w 8” option):</p>
<blockquote>
<div><ul class="simple">
<li><p>In “ordered” mode:</p></li>
</ul>
</div></blockquote>
</li>
</ol>
<p><code class="code docutils literal notranslate"><span class="pre">macsyfinder</span> <span class="pre">--sequence-db</span> <span class="pre">VICH001.B.00001.C001.fasta</span> <span class="pre">-o</span> <span class="pre">macsyfinder_TXSScan_VICH001_ordered</span> <span class="pre">--models</span> <span class="pre">TXSScan</span> <span class="pre">all</span> <span class="pre">--db-type</span> <span class="pre">ordered_replicon</span> <span class="pre">-w</span> <span class="pre">8</span></code>
<cite># specified output folder: macsyfinder_TXSScan_VICH001_ordered</cite></p>
<blockquote>
<div><ul class="simple">
<li><p>In “unordered” mode:</p></li>
</ul>
</div></blockquote>
<p><code class="code docutils literal notranslate"><span class="pre">macsyfinder</span> <span class="pre">--sequence-db</span> <span class="pre">VICH001.B.00001.C001.fasta</span> <span class="pre">-o</span> <span class="pre">macsyfinder_TXSScan_VICH001_unordered</span> <span class="pre">--models</span> <span class="pre">TXSScan</span> <span class="pre">all</span> <span class="pre">--db-type</span> <span class="pre">unordered</span> <span class="pre">-w</span> <span class="pre">8</span></code>
<cite># specified output folder: macsyfinder_TXSScan_VICH001_unordered</cite></p>
<p>The documentation on the generated output files can be consulted <a class="reference internal" href="outputs.html#outputs"><span class="std std-ref">here</span></a>.
See also our FAQ: <a class="reference internal" href="FAQ.html#faq-search-mode"><span class="std std-ref">What search mode to be used?</span></a></p>
</div></blockquote>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>A more comprehensive example of genome datasets with dedicated command lines and expected output files
can be found <a class="reference external" href="https://doi.org/10.6084/m9.figshare.21716426.v1">here.</a></p>
</div>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="input.html" class="btn btn-neutral float-right" title="Input and Options of MacSyFinder" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2014-2025, &#39;Institut Pasteur (Paris), CNRS&#39;.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
  

<div class="licensing">
  <p>
    This documentation is licensed under <a href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License
      <img src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png">.
    </a>
  </p>
</div>


</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

<a class="github" href="https://github.com/gem-pasteur/macsyfinder">
    <img style="position: absolute; top: 0; right: 0; border: 0;"
         src="https://raw.githubusercontent.com/gem-pasteur/macsyfinder/refs/heads/main/doc/source/_static/github-banner.svg"
         alt="Fork me on GitHub" />
</a>



</body>
</html>