// @jupyterlab/testing/lib/index.js
"use strict";
/* -----------------------------------------------------------------------------
| Copyright (c) Jupyter Development Team.
| Distributed under the terms of the Modified BSD License.
|----------------------------------------------------------------------------*/
/**
 * @packageDocumentation
 * @module testing
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.JupyterServer = exports.simulate = exports.dismissDialog = exports.dangerDialog = exports.acceptDialog = exports.waitForDialog = exports.sleep = exports.framePromise = exports.isFulfilled = exports.signalToPromise = exports.signalToPromises = exports.expectFailure = exports.testEmission = void 0;
var common_1 = require("./common");
Object.defineProperty(exports, "testEmission", { enumerable: true, get: function () { return common_1.testEmission; } });
Object.defineProperty(exports, "expectFailure", { enumerable: true, get: function () { return common_1.expectFailure; } });
Object.defineProperty(exports, "signalToPromises", { enumerable: true, get: function () { return common_1.signalToPromises; } });
Object.defineProperty(exports, "signalToPromise", { enumerable: true, get: function () { return common_1.signalToPromise; } });
Object.defineProperty(exports, "isFulfilled", { enumerable: true, get: function () { return common_1.isFulfilled; } });
Object.defineProperty(exports, "framePromise", { enumerable: true, get: function () { return common_1.framePromise; } });
Object.defineProperty(exports, "sleep", { enumerable: true, get: function () { return common_1.sleep; } });
Object.defineProperty(exports, "waitForDialog", { enumerable: true, get: function () { return common_1.waitForDialog; } });
Object.defineProperty(exports, "acceptDialog", { enumerable: true, get: function () { return common_1.acceptDialog; } });
Object.defineProperty(exports, "dangerDialog", { enumerable: true, get: function () { return common_1.dangerDialog; } });
Object.defineProperty(exports, "dismissDialog", { enumerable: true, get: function () { return common_1.dismissDialog; } });
Object.defineProperty(exports, "simulate", { enumerable: true, get: function () { return common_1.simulate; } });
var start_jupyter_server_1 = require("./start_jupyter_server");
Object.defineProperty(exports, "JupyterServer", { enumerable: true, get: function () { return start_jupyter_server_1.JupyterServer; } });
